src/main/java/net/fabricmc/loom/util/enumwidener/EnumWidenerClassNode.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * This file is part of fabric-loom, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) 2016, 2017, 2018 FabricMC
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+package net.fabricmc.loom.util.enumwidener;
+
+import org.objectweb.asm.FieldVisitor;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.tree.ClassNode;
+
+public class EnumWidenerClassNode extends ClassNode {
```

Comment:
Class visitor?

Reply:
no, I think he's trying to save on object here. kindof a microoptimization but whatever

Reply:
You'd still have to allocate the classreader and classwriter though. Just the visitor would be lower overhead.

Reply:
I had a ClassVisitor earlier but I needed a ClassNode in order to shift the parameter annotations due to the 2 additional parameters in constructor signatures.

Reply:
What was wrong with just returning a `MethodVisitor` which added two to `visitParameterAnnotation`'s `parameter`s when the method is `<init>`?

Reply:
> What was wrong with just returning a `MethodVisitor` which added two to `visitParameterAnnotation`'s `parameter`s when the method is `<init>`?

Nothing; I had not thought that it would be possible. It should be better now.